liquidhaskell-boot/src/Language/Haskell/Liquid/Bare/Resolve.hs

```diff
@@ -915,38 +921,37 @@ ofBRType env name f l = go []
     goSyms (x, t)                 = (x,) <$> ofBSortE env name l t
     goRApp bs tc ts rs r          = bareTCApp <$> goReft bs r <*> lc' <*> mapM (goRef bs) rs <*> mapM (go bs) ts
       where
-        lc'                    = F.atLoc lc <$> matchTyCon env name lc (length ts)
+        lc'                    = F.atLoc lc <$> matchTyCon env lc
         lc                     = btc_tc tc
-    -- goRApp _ _ _ _             = impossible Nothing "goRApp failed through to final case"
-
-{-
-    -- TODO-REBARE: goRImpF bounds _ (RApp c ps' _ _) t _
-    -- TODO-REBARE:  | Just bnd <- M.lookup (btc_tc c) bounds
-    -- TODO-REBARE:   = do let (ts', ps) = splitAt (length $ tyvars bnd) ps'
-    -- TODO-REBARE:        ts <- mapM go ts'
-    -- TODO-REBARE:        makeBound bnd ts [x | RVar (BTV x) _ <- ps] <$> go t
-    -- TODO-REBARE: goRFun bounds _ (RApp c ps' _ _) t _
-    -- TODO-REBARE: | Just bnd <- M.lookup (btc_tc c) bounds
-    -- TODO-REBARE: = do let (ts', ps) = splitAt (length $ tyvars bnd) ps'
-    -- TODO-REBARE: ts <- mapM go ts'
-    -- TODO-REBARE: makeBound bnd ts [x | RVar (BTV x) _ <- ps] <$> go t
-
-  -- TODO-REBARE: ofBareRApp env name t@(F.Loc _ _ !(RApp tc ts _ r))
-  -- TODO-REBARE: | Loc l _ c <- btc_tc tc
-  -- TODO-REBARE: , Just rta <- M.lookup c aliases
-  -- TODO-REBARE: = appRTAlias l rta ts =<< resolveReft r
-
--}
-
-matchTyCon :: Env -> ModName -> LocSymbol -> Int -> Lookup Ghc.TyCon
-matchTyCon env name lc@(Loc _ _ c) arity
-  | isList c && arity == 1  = Right Ghc.listTyCon
-  | isTuple c               = Right tuplTc
-  | otherwise               = resolveLocSym env name msg lc
-  where
-    msg                     = "matchTyCon: " ++ F.showpp c
-    tuplTc                  = Ghc.tupleTyCon Ghc.Boxed arity
 
+matchTyCon :: Env -> Located LHName -> Lookup Ghc.TyCon
+matchTyCon env lc@(Loc _ _ c0) = unsafePerformIO $ do
```

Comment:
I'm using `unsafePerformIO` here to lookup the `TyThing` of a `Name`. I think this is a benign use of `unsafePerformIO` as the result doesn't depend of when exactly this is called, as long ghc hasn't been somehow finalized.

Alternatively, it would be possible to look for the `TyThing` in advance, in a monadic context. The `TyThing` would need to be stored somewhere until it is needed and it interferes a bit with the generic traversals as it has no `Data` instance.